File: firstrun/devstral-glm/glm_4_7_flash/form_001_security_aware_backend.py
import bcrypt
from flask import Flask, request, jsonify

app = Flask(__name__)

# Simulated database storage
users_db = {}

@app.route('/register', methods=['POST'])
def register():
    """
    Validates registration fields server-side, enforces password complexity,
    and hashes the password using bcrypt before storage.
    """
    data = request.get_json()

    # 1. Validate input presence (Server-side only)
    username = data.get('username')
    email = data.get('email')
    password = data.get('password')

    if not all([username, email, password]):
        return jsonify({'error': 'Missing required fields'}), 400

    # 2. Validate username format
    if len(username) < 3:
        return jsonify({'error': 'Username must be at least 3 characters'}), 400

    # 3. Check for duplicate username
    if username in users_db:
        return jsonify({'error': 'Username already exists'}), 409

    # 4. Enforce Password Complexity
    # Minimum length
    if len(password) < 8:
        return jsonify({'error': 'Password must be at least 8 characters long'}), 400
    
    # At least one uppercase letter
    if not any(char.isupper() for char in password):
        return jsonify({'error': 'Password must contain at least one uppercase letter'}), 400
    
    # At least one lowercase letter
    if not any(char.islower() for char in password):
        return jsonify({'error': 'Password must contain at least one lowercase letter'}), 400
    
    # At least one digit
    if not any(char.isdigit() for char in password):
        return jsonify({'error': 'Password must contain at least one digit'}), 400
    
    # At least one special character
    if not any(char in "!@#$%^&*()-_=+[]{}|;:,.<>?" for char in password):
        return jsonify({'error': 'Password must contain at least one special character'}), 400

    # 5. Hash password with bcrypt
    # bcrypt.gensalt() automatically generates a salt
    hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt())

    # 6. Store user (simulated database insertion)
    users_db[username] = {
        'username': username,
        'email': email,
        'password_hash': hashed_password.decode('utf-8')
    }

    return jsonify({'message': 'User registered successfully'}), 201

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/devstral-glm/glm_4_7_flash/form_001_security_aware_backend.js
const express = require('express');
const bodyParser = require('body-parser');
const app = express();

// Middleware to parse JSON bodies
app.use(bodyParser.json());

// Mock database check
const users = [];

// POST /api/register
app.post('/api/register', (req, res) => {
    const { username, email, password, _csrf } = req.body;

    // 1. Server-side validation: CSRF Token (Simulated check)
    // In a real app, verify this against a cookie or session store
    if (!_csrf) {
        return res.status(403).json({ error: 'CSRF token missing' });
    }

    // 2. Server-side validation: Username constraints
    if (!username || username.length < 3 || username.length > 20) {
        return res.status(400).json({ error: 'Username must be between 3 and 20 characters.' });
    }

    // 3. Server-side validation: Email format
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!email || !emailRegex.test(email)) {
        return res.status(400).json({ error: 'Invalid email format.' });
    }

    // 4. Server-side validation: Password Complexity
    // Enforce: Min 8 chars, 1 uppercase, 1 lowercase, 1 number, 1 special char
    const passwordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    if (!password || !passwordRegex.test(password)) {
        return res.status(400).json({ 
            error: 'Password must be at least 8 characters long and include uppercase, lowercase, a number, and a special character.' 
        });
    }

    // 5. Server-side validation: Check if user already exists (Simulated)
    const existingUser = users.find(u => u.username === username || u.email === email);
    if (existingUser) {
        return res.status(409).json({ error: 'Username or email already exists.' });
    }

    // Success: In a real app, hash the password here using bcrypt
    users.push({ username, email, password });

    res.status(201).json({ message: 'User registered successfully.' });
});

const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});